pkg/remote/write.go

```diff
@@ -0,0 +1,253 @@
+	"github.com/grafana/prometheus-pulsar-remote-write/pkg/pulsar"
+	"github.com/grafana/prometheus-pulsar-remote-write/pkg/version"
+)
+
+type Write struct {
+	logger log.Logger
+
+	samplePerTenantID   map[string][]pulsar.ReceivedSample
+	deadlinePerTenantID map[string]time.Time
+
+	// BatchSize is the amount of samples that are aggregated into a single
+	// remote write request
+	BatchSize int
+	// BatchMaxDelay is the maximum delay acceptable for a batch to wait to
+	// reach the BatchSize
+	BatchMaxDelay time.Duration
+
+	// checkInterval limits how often we want to check on send conditions for batches are met
+	checkInterval time.Duration
+}
+
+type WriteOpts func(o *Write)
+
+func WithLogger(l log.Logger) WriteOpts {
+	return func(w *Write) {
+		w.logger = l
+	}
+}
+
+func NewWrite(opts ...WriteOpts) *Write {
+	w := &Write{
+		logger: log.NewNopLogger(),
+
+		samplePerTenantID:   make(map[string][]pulsar.ReceivedSample),
+		deadlinePerTenantID: make(map[string]time.Time),
+
+		BatchSize:     100,
+		BatchMaxDelay: 5 * time.Second,
+		checkInterval: 100 * time.Millisecond,
+	}
+
+	for _, opt := range opts {
+		opt(w)
+	}
+
+	return w
+}
+
+type headerRoundtripper struct {
+	upstream http.RoundTripper
+}
+
+func (t *headerRoundtripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	// if tenant ID is set, expose it in the header
+	tenantID := mcontext.TenantIDFromContext(req.Context())
+	if tenantID != "" {
+		req.Header.Set(mcontext.HTTPHeaderTenantID, tenantID)
+	}
+
+	// overwrite the user agent header
+	req.Header.Set("User-Agent", fmt.Sprintf("%s/%s", version.AppName(), version.AppVersion()))
+
+	return t.upstream.RoundTrip(req)
+}
+
+type ClientConfig = remote.ClientConfig
+
+func NewWriteClient(conf *ClientConfig) (remote.WriteClient, error) {
+	clientInterface, err := remote.NewWriteClient("remote-write", conf)
+	if err != nil {
+		return nil, err
+	}
+
+	client, ok := clientInterface.(*remote.Client)
+	if !ok {
+		return nil, fmt.Errorf("client of unexpected type %T", clientInterface)
+	}
+
+	// add custom rountripper to modify user-agent and tenant header if necessary
+	client.Client.Transport = &headerRoundtripper{upstream: client.Client.Transport}
+
+	return client, nil
+}
+
+func (w *Write) Run(ctx context.Context, sampleCh chan pulsar.ReceivedSample, client remote.WriteClient) error {
+	tick := time.NewTicker(w.checkInterval)
+	defer tick.Stop()
+
+	// this is set when a retry able error happend
+	errRemoteWriteRetryable := false
+	blockingSampleCh := make(chan pulsar.ReceivedSample)
+
+	sample := func() chan pulsar.ReceivedSample {
+		// don't return new samples if err has happend
+		if errRemoteWriteRetryable {
+			return blockingSampleCh
+		}
+		return sampleCh
+	}
+
+	var samplesSinceCheck int
+	var deadlineCheck time.Time
+
+receive:
+
+	for {
+		// wait for either tick or incoming sample, handle closed context
+		select {
+		case <-ctx.Done():
+			break receive
```

Comment:
I don't understand how this `break receive` works? It just restarts the loop. `ctx.Done()` will still return the same value. So won't it just loop?

Reply:
Not too sure if I am missing something, but I think its quite similar to the behaviour here: https://www.ardanlabs.com/blog/2013/11/label-breaks-in-go.html

To avoid "breaking" the select, I put a label before the for loop and breaks the for loop after the label `receive`. I have removed the new line to make it more explicit.